src/parquet/file/metadata-internal.cc

```diff
@@ -0,0 +1,525 @@
+  return impl_->IsStatsSet();
+}
+
+Compression::type ColumnMetaData::GetCompression() const {
+  return impl_->GetCompression();
+}
+
+std::vector<Encoding::type> ColumnMetaData::GetEncodings() const {
+  return impl_->GetEncodings();
+}
+
+int64_t ColumnMetaData::GetUnCompressedSize() const {
+  return impl_->GetUnCompressedSize();
+}
+
+int64_t ColumnMetaData::GetCompressedSize() const {
+  return impl_->GetCompressedSize();
+}
+
+// row-group metadata
+class RowGroupMetaData::RowGroupMetaDataImpl {
+ public:
+  explicit RowGroupMetaDataImpl(const format::RowGroup* row_group)
+      : row_group_(row_group) {}
+  ~RowGroupMetaDataImpl() {}
+
+  int num_columns() const { return row_group_->columns.size(); }
+
+  int64_t num_rows() const { return row_group_->num_rows; }
+
+  int64_t total_byte_size() const { return row_group_->total_byte_size; }
+
+  std::unique_ptr<ColumnMetaData> GetColumnMetaData(int i) {
+    DCHECK(i < num_columns()) << "The file only has " << num_columns()
+                              << " columns, requested metadata for column: " << i;
+    return ColumnMetaData::GetColumnMetaData(
+        reinterpret_cast<const uint8_t*>(&row_group_->columns[i]));
+  }
+
+ private:
+  const format::RowGroup* row_group_;
+};
+
+std::unique_ptr<RowGroupMetaData> RowGroupMetaData::GetRowGroupMetaData(
+    const uint8_t* metadata) {
+  return std::unique_ptr<RowGroupMetaData>(new RowGroupMetaData(metadata));
+}
+
+RowGroupMetaData::RowGroupMetaData(const uint8_t* metadata)
+    : impl_{std::unique_ptr<RowGroupMetaDataImpl>(new RowGroupMetaDataImpl(
+          reinterpret_cast<const format::RowGroup*>(metadata)))} {}
+RowGroupMetaData::~RowGroupMetaData() {}
+
+int RowGroupMetaData::num_columns() const {
+  return impl_->num_columns();
+}
+
+int64_t RowGroupMetaData::num_rows() const {
+  return impl_->num_rows();
+}
+
+int64_t RowGroupMetaData::total_byte_size() const {
+  return impl_->total_byte_size();
+}
+
+std::unique_ptr<ColumnMetaData> RowGroupMetaData::GetColumnMetaData(int i) const {
+  return impl_->GetColumnMetaData(i);
+}
+
+// file metadata
+class FileMetaData::FileMetaDataImpl {
+ public:
+  explicit FileMetaDataImpl(const uint8_t* metadata)
+      : metadata_{std::unique_ptr<const format::FileMetaData>(
+            reinterpret_cast<const format::FileMetaData*>(metadata))} {
+    schema::FlatSchemaConverter converter(
+        &metadata_->schema[0], metadata_->schema.size());
+    schema_.Init(converter.Convert());
+  }
+  ~FileMetaDataImpl() {}
+  int64_t num_rows() const { return metadata_->num_rows; }
+  int num_row_groups() const { return metadata_->row_groups.size(); }
+  int32_t version() const { return metadata_->version; }
+  const std::string& created_by() const { return metadata_->created_by; }
+  int num_schema_elements() const { return metadata_->schema.size(); }
+
+  void WriteTo(OutputStream* dst) {
+    SerializeThriftMsg(metadata_.get(), sizeof(format::FileMetaData), dst);
```

Comment:
The way we are using `SerializeThriftMsg` is a bit odd as the 2nd argument is supposed (from what I can tell) to be an estimated size of the serialized object. I created PARQUET-690 to look into this separately